Sprint 20) Перевод поиска по iTunes, истории поиска и debounce на корутины

amdoit-git opened the pull request:
Что предстоит сделать:

1. Переделайте реализацию debounce для отправки поискового запроса и нажатия на элементы списка. Вместо прежнего подхода с использованием Handler используйте корутины. Это позволит сделать процесс более эффективным и управляемым.
2. Обновите реализацию отображения прогресса воспроизведения на экране «Аудиоплеер». Теперь в этой реализации нужно использовать корутины. Это обеспечит более простое и надёжное отображение прогресса.
3. Наконец, выполнение самого сетевого запроса для поиска треков теперь должно быть реализовано с использованием корутин и Flow на всех уровнях архитектуры исходного кода. Это позволит вам дополнительно попрактиковаться в использовании этих инструментов в полном масштабе и на конкретной задаче. Вы получите возможность глубже понять, как они обеспечивают выполнение асинхронных операций и помогают работать с потоками данных.

amdoit-git committed:
Отображения прогресса воспроизведения на экране «Аудиоплеер» переделано с handler на корутины

amdoit-git committed:
Реализацию debounce для отправки поискового запроса и нажатия на элементы списка переделана на корутины

amdoit-git committed:
Выполнение сетевого запроса для поиска треков реализовано с использованием корутин и Flow

amdoit-git committed:
Выполнение сохранения/чтения истории поиска треков и сохраниния/чтения настроект темы приложения реализовано с использованием корутин

amdoit-git committed:
Удалены строчки с логированием

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/player/MediaPlayerService.kt:
```diff
@@ -119,22 +123,31 @@ class MediaPlayerService(private val mediaPlayer: MediaPlayer) {
+    private fun startTimer() {
+
+        if (timerJob == null) {
+            timerJob = GlobalScope.launch(Dispatchers.Main) {
```
Очень плохо использовать GlobalScope. Нужно либо делать кастомный и следить за жизненным циклом, либо как-то подвязывать к ЖЦ Андроида.

amdoit-git replied:
> Очень плохо использовать GlobalScope. Нужно либо делать кастомный и следить за жизненным циклом, либо как-то подвязывать к ЖЦ Андроида.
Вообще сам single<MediaPlayerService> тоже идет как сингелтон и может работать в фоне независимо от экрана приложения на котором он впервые был вызван.

kekulta replied:
В больших приложениях у тебя могут появляться такие штуки как DI скоупы. Т.е. синглом он будет являться только в определенной части приложения и тогда глобал скоуп может начать утекать.

Ну и опять же, есть просто хорошие практики, которых лучше придерживаться, чтобы потом случайно не отстрелить себе ногу.

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/player/MediaPlayerService.kt:
```diff
@@ -119,22 +123,31 @@ class MediaPlayerService(private val mediaPlayer: MediaPlayer) {
+                while (true) {
+                    displayPlayProgress()
+                    delay(1000L)
+                }
```
Должен обновляться раз в 300мс.

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/search/ITunesRepositoryImpl.kt:
```diff
@@ -3,153 +3,58 @@ package com.example.playlistmaker.data.impl.search
+            } catch (error: CancellationException) {
 
-            if (text != getSearchText()) {
-                return SearchState.NEW_SEARCH_NEEDED
-            }
+                ConsumerData.Error(
+                    code = -2,
+                    message = "The search was terminated"
+                )
```
Очень мощное нарушение ЖЦ корутин, CancellationException нужно пробрасывать дальше, иначе не получится отменить корутину.

amdoit-git replied:
В смысле не получится отменить? Там же ниже нет никакого кода. У корутины тут нет вариантов кроме как завершится

kekulta replied:
Есть такая штука как ментальная модель. Когда мы отменяем корутину, то мы, как разработчики ждём определенного поведения. В частности, это проброс CancellationException и отмена работы и тихая смерть корутины. В данном случае, ты ломаешь этот контракт и это не очень хорошо, особенно в больших проектах.

Это примерно как с кастомной логикой в hashcode и equals. Да, теоретически так можно сделать и можно даже получить некоторую лаконичность из такого хака, но чаще всего оно не стоит сломанного контракта и шанса, что другие разработчики запутаются.

Да, в данном проекте ты один, но было бы уже сразу неплохо прививать хорошие привычки и не ломать общеязыковые контракты.

amdoit-git replied:
Понял. Сделал проброс исключения. Причем код даже стал проще. Избавился благодаря этому от ненужного типа в  ConsumerData. Спасибо))

kekulta commented on app/src/main/java/com/example/playlistmaker/ui/common/App.kt:
```diff
@@ -28,6 +32,8 @@ class App : Application() {
         val settings: AppSettingsInteractor by inject()
 
-        settings.restoreTheme()
+        runBlocking {
+            settings.restoreTheme()
+        }
```
Весь смысл корутин теряется, если запускать runBlocking.

amdoit-git replied:
> Весь смысл корутин теряется, если запускать runBlocking.

Это специально было сделано. При асинхронном запуске получается приложение открывается. И пока идет чтение настроек успевает нарисовать экран с темой по умолчанию и потом если тема темная видно мерцание. А если читать как и было синхронно в главном потоке, то он на это время блокируется и ничего не рисует. Так что блокировку специально оставил

kekulta replied:
Валидный поинт, вопрос снимается.

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/player/MediaPlayerService.kt:
```diff
@@ -29,8 +34,7 @@ class MediaPlayerService(private val mediaPlayer: MediaPlayer) {
-    private val obj = Any()
+    private var timerJob: Job? = null
 
     fun setDisplayPorts(
         forTime: ((Int) -> Unit)?,
```
Будет правильнее сделать на корутинах.

amdoit-git committed:
Прогреcc воспроизведения обновляется раз в 300мс.
Функции интерактора не возвращают Flow.
Сделал проброс исключения CancellationException

amdoit-git committed:
Мало ли что имеется в виду под фразой "возвращают Flow". Решил, что лучше перестраховаться и подписаться в интеракторе на flow репозитория и эмитировать его дальше. Чем просто вернуть flow репозитория.

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/search/ITunesRepositoryImpl.kt:
```diff
@@ -41,18 +48,18 @@ class ITunesRepositoryImpl(private val context: Context, private val api: Itunes
-                )
+                if(error is CancellationException){
+                    throw error
+                }
```
Можно сделать так, можно сделать как как ты делал до этого:
```
try {
   ...
} catch (e: CancellationException) {
    throw e
} catch (e: Exception) {
    ...
}
```

kekulta replied:
Обычно предпочитают вариант с двумя кетчами, потомучто вариант с if добавляет лишней уровень вложенности, а так же может быть дороже в плане перфоманса.

kekulta commented on app/src/main/java/com/example/playlistmaker/data/impl/search/ITunesRepositoryImpl.kt:
```diff
@@ -41,18 +48,18 @@ class ITunesRepositoryImpl(private val context: Context, private val api: Itunes
-                )
-            } catch (error: Throwable) {
+            }
+            catch (error: Throwable) {
```
Плохая практика. У тебя подклассом Throwable является не только Exception, но ещё и Error. 
Error ловить не нужно, хотя и физически возможно. Это ошибки из которых нельзя восстановится, например,OutOfMemory или StackOverflow. Нужно кэтчить Exception.

kekulta commented on app/src/main/java/com/example/playlistmaker/domain/impl/search/ITunesInteractorImpl.kt:
```diff
@@ -11,6 +11,6 @@ class ITunesInteractorImpl(private val repository: ITunesRepository) : ITunesInt
     override suspend fun search(text: String): Flow<ConsumerData<List<Track>>> = flow {
 
-        emit(repository.search(text))
+        repository.search(text).collect { data -> emit(data) }
```
Можно просто возврашать флоу из репозитория.

amdoit-git committed:
Исправления по замечаниям ревьювера
